Adding support for Google Cloud Storage as a storage provider

gauntface opened the pull request:
After talking with @jeffmcaffer a little last week this adds support for Google Cloud Storage to GHCrawler (Both as a storage provider and a delta provider).

I've tried to mimic the azure storage model as much as possible, couple of things I'm not sure of:

1. I couldn't get the delta options to get passed from `initializeSubsystemOptions()` in the crawlerFactory. The `CRAWLER_DELTA_PROVIDER` value was being picked up by the `getDefaultOptions()` method, but would get filtered out by `initializeSubsystemOptions()`. Commenting out the early return fixes this but I'm sure I'm missing the purpose of that if statement.
1. I've used regular promises instead of the Q library promises, is this ok?
1. The approach taken in the delta store is to write the contents to unique files in google cloud storage as I don't believe this is an efficient way to append to a file (which I think there is in the azure storage module).

Happy to change any of the approaches, code style etc, just let me know where the issues are :)

jeffmcaffer commented on lib/crawlerFactory.js:
```diff
@@ -206,17 +208,18 @@ class CrawlerFactory {
-    if (Object.getOwnPropertyNames(config).length > 1) {
-      return Q(config);
-    }
+    // TODO: This prevents delta values being propagated throughout the factory.
```
Did you previously run the crawler using a redis (or other persistent) config store? The idea here is that if the `getAll` in the caller (line 205) gets some values (i.e., there is more than one prop) then we should use them. This is setup to share a config across many instances. If you want to change the config you have to either change them in the store (e.g., redis), clear the store (e.g., redis) or do some crawler API or crawler dashboard calls to change. you can see the current config by starting up the crawler and use the [crawler cli](https://github.com/Microsoft/ghcrawler-cli).

Here is the comment that should appear at line 217 in the original.
```javascript
    // if the incoming config already has values (e.g., retrieved from Redis) then assume it already initialized
    // and do not apply defaults. 
    // TODO: consider a mechanism for forcing flushing config. Would need to ensure this sync'd up well with whatever
    // the persistent config provider is doing.
    // TODO: here we are relying on an implementation detail of the config object that injects a `_config` prop. We 
    // should update the config code to make that prop non enumerable so we don't have to filter it out everywhere.
```

If that is not the problem, can you attach your `env.json` (with the secrets obfuscated) so we can repro?

gauntface replied:
I honestly have no idea what the problem was that I was seeing. I've added the comment to the code.

Originally the defaults were getting swallowed by a config else where, but now it seems fine - so I'm going to assume this was a docker caching issue somewhere and step away.

jeffmcaffer commented on providers/storage/googleCloudDeltaStorage.js:
```diff
@@ -0,0 +1,106 @@
+  _stashDelta(text) {
+    const fileName = this._getFilename(text);
+
+    this.deltaSequenceNumber++;
```
This is only ever used here

jeffmcaffer commented on providers/storage/googleCloudDeltaStorage.js:
```diff
@@ -0,0 +1,106 @@
+    const formattedDate = now.format('YYYY_MM_DD_HH');
+
+    if (hour !== this.currentHour) {
+      this.currentHour = hour;
```
`currentHour` does not seem to be used for much

jeffmcaffer commented on lib/crawlerFactory.js:
```diff
@@ -230,7 +233,7 @@ class CrawlerFactory {
     const config = new RefreshingConfig.RefreshingConfig(configStore)
       .withExtension(new RefreshingConfigRedis.RedisPubSubRefreshPolicyAndChangePublisher(redisClient, channel));
-    return config;
+      return config;
```
extra indent not needed

jeffmcaffer commented on providers/storage/googleCloudDeltaStorage.js:
```diff
@@ -0,0 +1,106 @@
+
+  connect() {
+    return this.baseStore.connect()
+    .then(() => {
```
formatting nit: General indenting pattern in the rest of the code is to have the `then` a level below the receiver. (so in a level on this line)

There are a few spots like this

gauntface replied:
Updated in the delta and storage files.

gauntface commented:
@jeffmcaffer thanks for the review comments.

I've switched everything to Q where I was originally creating native promises. I haven't gone to the extent of wrapping the promises returned by the Google Cloud Storage package, let me know if you would prefer it though.

Added some tests for the new delta store - I didn't see a coverage reporting anywhere, I've just been running `npm run test` to check everything is WAI.

Let me know if you'd like any other changes / tests / etc. 

Regarding testing - I am running this on a container to make sure it's crawling and working as I'd expect so far, but obviously doesn't rule out stupid mistakes on my part, so feel free to question any of this PR.

gauntface commented:
@jeffmcaffer I'm starting to see the original problem crop up with the config, the deadletter config has begun to fail for me - will dig in and update shortly.
